<a href="https://colab.research.google.com/github/anabiarochar/Mestrado_AlgoritmosProgramacao/blob/main/Projeto1/5_QuickSort.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

##Projeto 1 - Algoritmo de Ordenação Quick Sort

Abaixo estão os algorítmos para ordenação, utilizando o método quick sort de ordenação, com entradas de vetores de tamanhos 1.000, 10.000 e 100.000

- aleatórias
- ordenadas em forma crescente
- ordenadas em forma decrescente  

Abaixo de cada algorítmo estão os tempos observados somando 4 tentativas.

##Vetor em ordem aleatória com 1000 elementos:

In [5]:
import requests
import time
import random

# Importar dados
url = "https://raw.githubusercontent.com/anabiarochar/Mestrado_AlgoritmosProgramacao/main/Projeto1/Vetores/vetor_1000_aleatorio.txt"

try:
    response = requests.get(url)
    content = response.text
    # Ler arquivo como vetor separado por vírgula
    vetor = [int(x.strip()) for x in content.split(',') if x.strip()]
except Exception as e:
    print(f"Erro ao carregar o arquivo: {e}")

# Método QuickSort com pivô aleatório
def quick_sort_random(arr):
    if len(arr) <= 1:
        return arr

    indice_pivo = random.randint(0, len(arr) - 1)
    pivo = arr[indice_pivo]

    esquerda = [x for i, x in enumerate(arr) if x < pivo]
    meio = [x for x in arr if x == pivo]
    direita = [x for i, x in enumerate(arr) if x > pivo]

    return quick_sort_random(esquerda) + meio + quick_sort_random(direita)

# Execução e Tempo
if vetor:

    inicio = time.perf_counter()
    vetor_ordenado = quick_sort_random(vetor)
    fim = time.perf_counter()

    tempo_total = fim - inicio

    print(f"Tempo de execução: {tempo_total:.6f} segundos")
    print(vetor_ordenado[:10])
else:
    print("Não foi possível processar o vetor.")

Tempo de execução: 0.002113 segundos
[1, 3, 8, 9, 9, 10, 10, 10, 10, 12]


###Tempo de Execução:

Tentativa 1 - Tempo de execução: 0.003436 segundos  
Tentativa 2 - Tempo de execução: 0.002381 segundos  
Tentativa 3 - Tempo de execução: 0.002268 segundos  
Tentativa 4 - Tempo de execução: 0.002113 segundos  

##Vetor em ordem aleatória com 10000 elementos:

In [9]:
import requests
import time
import random

# Importar dados
url = "https://raw.githubusercontent.com/anabiarochar/Mestrado_AlgoritmosProgramacao/main/Projeto1/Vetores/vetor_10000_aleatorio.txt"

try:
    response = requests.get(url)
    content = response.text
    # Ler arquivo como vetor separado por vírgula
    vetor = [int(x.strip()) for x in content.split(',') if x.strip()]
except Exception as e:
    print(f"Erro ao carregar o arquivo: {e}")

# Método QuickSort com pivô aleatório
def quick_sort_random(arr):
    if len(arr) <= 1:
        return arr

    indice_pivo = random.randint(0, len(arr) - 1)
    pivo = arr[indice_pivo]

    esquerda = [x for i, x in enumerate(arr) if x < pivo]
    meio = [x for x in arr if x == pivo]
    direita = [x for i, x in enumerate(arr) if x > pivo]

    return quick_sort_random(esquerda) + meio + quick_sort_random(direita)

# Execução e Tempo
if vetor:

    inicio = time.perf_counter()
    vetor_ordenado = quick_sort_random(vetor)
    fim = time.perf_counter()

    tempo_total = fim - inicio

    print(f"Tempo de execução: {tempo_total:.6f} segundos")
    print(vetor_ordenado[:10])
else:
    print("Não foi possível processar o vetor.")

Tempo de execução: 0.032041 segundos
[3, 3, 4, 8, 9, 9, 9, 11, 12, 12]


###Tempo de Execução:

Tentativa 1 - Tempo de execução: 0.030176 segundos  
Tentativa 2 - Tempo de execução: 0.064724 segundos  
Tentativa 3 - Tempo de execução: 0.034632 segundos  
Tentativa 4 - Tempo de execução: 0.032041 segundos  

##Vetor em ordem aleatória com 100000 elementos:

In [14]:
import requests
import time
import random

# Importar dados
url = "https://raw.githubusercontent.com/anabiarochar/Mestrado_AlgoritmosProgramacao/main/Projeto1/Vetores/vetor_100000_aleatorio.txt"

try:
    response = requests.get(url)
    content = response.text
    # Ler arquivo como vetor separado por vírgula
    vetor = [int(x.strip()) for x in content.split(',') if x.strip()]
except Exception as e:
    print(f"Erro ao carregar o arquivo: {e}")

# Método QuickSort com pivô aleatório
def quick_sort_random(arr):
    if len(arr) <= 1:
        return arr

    indice_pivo = random.randint(0, len(arr) - 1)
    pivo = arr[indice_pivo]

    esquerda = [x for i, x in enumerate(arr) if x < pivo]
    meio = [x for x in arr if x == pivo]
    direita = [x for i, x in enumerate(arr) if x > pivo]

    return quick_sort_random(esquerda) + meio + quick_sort_random(direita)

# Execução e Tempo
if vetor:

    inicio = time.perf_counter()
    vetor_ordenado = quick_sort_random(vetor)
    fim = time.perf_counter()

    tempo_total = fim - inicio

    print(f"Tempo de execução: {tempo_total:.6f} segundos")
    print(vetor_ordenado[:10])
else:
    print("Não foi possível processar o vetor.")

Tempo de execução: 0.452698 segundos
[2, 5, 6, 6, 8, 9, 10, 13, 14, 14]


###Tempo de Execução:

Tentativa 1 - Tempo de execução: 0.860136 segundos  
Tentativa 2 - Tempo de execução: 0.465810 segundos  
Tentativa 3 - Tempo de execução: 0.523734 segundos  
Tentativa 4 - Tempo de execução: 0.452698 segundos  

##Vetor em ordem crescente com 1000 elementos:

In [18]:
import requests
import time
import random

# Importar dados
url = "https://raw.githubusercontent.com/anabiarochar/Mestrado_AlgoritmosProgramacao/main/Projeto1/Vetores/vetor_1000_crescente.txt"

try:
    response = requests.get(url)
    content = response.text
    # Ler arquivo como vetor separado por vírgula
    vetor = [int(x.strip()) for x in content.split(',') if x.strip()]
except Exception as e:
    print(f"Erro ao carregar o arquivo: {e}")

# Método QuickSort com pivô aleatório
def quick_sort_random(arr):
    if len(arr) <= 1:
        return arr

    indice_pivo = random.randint(0, len(arr) - 1)
    pivo = arr[indice_pivo]

    esquerda = [x for i, x in enumerate(arr) if x < pivo]
    meio = [x for x in arr if x == pivo]
    direita = [x for i, x in enumerate(arr) if x > pivo]

    return quick_sort_random(esquerda) + meio + quick_sort_random(direita)

# Execução e Tempo
if vetor:

    inicio = time.perf_counter()
    vetor_ordenado = quick_sort_random(vetor)
    fim = time.perf_counter()

    tempo_total = fim - inicio

    print(f"Tempo de execução: {tempo_total:.6f} segundos")
    print(vetor_ordenado[:10])
else:
    print("Não foi possível processar o vetor.")

Tempo de execução: 0.007712 segundos
[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]


###Tempo de Execução:

Tentativa 1 - Tempo de execução: 0.007867 segundos  
Tentativa 2 - Tempo de execução: 0.003974 segundos  
Tentativa 3 - Tempo de execução: 0.002431 segundos  
Tentativa 4 - Tempo de execução: 0.007712 segundos  

##Vetor em ordem crescente com 10000 elementos:

In [22]:
import requests
import time
import random

# Importar dados
url = "https://raw.githubusercontent.com/anabiarochar/Mestrado_AlgoritmosProgramacao/main/Projeto1/Vetores/vetor_10000_crescente.txt"

try:
    response = requests.get(url)
    content = response.text
    # Ler arquivo como vetor separado por vírgula
    vetor = [int(x.strip()) for x in content.split(',') if x.strip()]
except Exception as e:
    print(f"Erro ao carregar o arquivo: {e}")

# Método QuickSort com pivô aleatório
def quick_sort_random(arr):
    if len(arr) <= 1:
        return arr

    indice_pivo = random.randint(0, len(arr) - 1)
    pivo = arr[indice_pivo]

    esquerda = [x for i, x in enumerate(arr) if x < pivo]
    meio = [x for x in arr if x == pivo]
    direita = [x for i, x in enumerate(arr) if x > pivo]

    return quick_sort_random(esquerda) + meio + quick_sort_random(direita)

# Execução e Tempo
if vetor:

    inicio = time.perf_counter()
    vetor_ordenado = quick_sort_random(vetor)
    fim = time.perf_counter()

    tempo_total = fim - inicio

    print(f"Tempo de execução: {tempo_total:.6f} segundos")
    print(vetor_ordenado[:10])
else:
    print("Não foi possível processar o vetor.")

Tempo de execução: 0.059693 segundos
[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]


###Tempo de Execução:

Tentativa 1 - Tempo de execução: 0.030033 segundos  
Tentativa 2 - Tempo de execução: 0.031734 segundos  
Tentativa 3 - Tempo de execução: 0.032171 segundos  
Tentativa 4 - Tempo de execução: 0.059693 segundos  

##Vetor em ordem crescente com 100000 elementos:

In [26]:
import requests
import time
import random

# Importar dados
url = "https://raw.githubusercontent.com/anabiarochar/Mestrado_AlgoritmosProgramacao/main/Projeto1/Vetores/vetor_100000_crescente.txt"

try:
    response = requests.get(url)
    content = response.text
    # Ler arquivo como vetor separado por vírgula
    vetor = [int(x.strip()) for x in content.split(',') if x.strip()]
except Exception as e:
    print(f"Erro ao carregar o arquivo: {e}")

# Método QuickSort com pivô aleatório
def quick_sort_random(arr):
    if len(arr) <= 1:
        return arr

    indice_pivo = random.randint(0, len(arr) - 1)
    pivo = arr[indice_pivo]

    esquerda = [x for i, x in enumerate(arr) if x < pivo]
    meio = [x for x in arr if x == pivo]
    direita = [x for i, x in enumerate(arr) if x > pivo]

    return quick_sort_random(esquerda) + meio + quick_sort_random(direita)

# Execução e Tempo
if vetor:

    inicio = time.perf_counter()
    vetor_ordenado = quick_sort_random(vetor)
    fim = time.perf_counter()

    tempo_total = fim - inicio

    print(f"Tempo de execução: {tempo_total:.6f} segundos")
    print(vetor_ordenado[:10])
else:
    print("Não foi possível processar o vetor.")

Tempo de execução: 0.425225 segundos
[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]


###Tempo de Execução:

Tentativa 1 - Tempo de execução: 0.366315 segundos  
Tentativa 2 - Tempo de execução: 0.385219 segundos  
Tentativa 3 - Tempo de execução: 0.404614 segundos  
Tentativa 4 - Tempo de execução: 0.425225 segundos  

##Vetor em ordem decrescente com 1000 elementos:

In [30]:
import requests
import time
import random

# Importar dados
url = "https://raw.githubusercontent.com/anabiarochar/Mestrado_AlgoritmosProgramacao/main/Projeto1/Vetores/vetor_1000_decrescente.txt"

try:
    response = requests.get(url)
    content = response.text
    # Ler arquivo como vetor separado por vírgula
    vetor = [int(x.strip()) for x in content.split(',') if x.strip()]
except Exception as e:
    print(f"Erro ao carregar o arquivo: {e}")

# Método QuickSort com pivô aleatório
def quick_sort_random(arr):
    if len(arr) <= 1:
        return arr

    indice_pivo = random.randint(0, len(arr) - 1)
    pivo = arr[indice_pivo]

    esquerda = [x for i, x in enumerate(arr) if x < pivo]
    meio = [x for x in arr if x == pivo]
    direita = [x for i, x in enumerate(arr) if x > pivo]

    return quick_sort_random(esquerda) + meio + quick_sort_random(direita)

# Execução e Tempo
if vetor:

    inicio = time.perf_counter()
    vetor_ordenado = quick_sort_random(vetor)
    fim = time.perf_counter()

    tempo_total = fim - inicio

    print(f"Tempo de execução: {tempo_total:.6f} segundos")
    print(vetor_ordenado[:10])
else:
    print("Não foi possível processar o vetor.")

Tempo de execução: 0.002422 segundos
[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]


###Tempo de Execução:

Tentativa 1 - Tempo de execução: 0.004115 segundos  
Tentativa 2 - Tempo de execução: 0.002471 segundos  
Tentativa 3 - Tempo de execução: 0.004370 segundos  
Tentativa 4 - Tempo de execução: 0.002422 segundos  

##Vetor em ordem decrescente com 10000 elementos:

In [35]:
import requests
import time
import random

# Importar dados
url = "https://raw.githubusercontent.com/anabiarochar/Mestrado_AlgoritmosProgramacao/main/Projeto1/Vetores/vetor_10000_decrescente.txt"

try:
    response = requests.get(url)
    content = response.text
    # Ler arquivo como vetor separado por vírgula
    vetor = [int(x.strip()) for x in content.split(',') if x.strip()]
except Exception as e:
    print(f"Erro ao carregar o arquivo: {e}")

# Método QuickSort com pivô aleatório
def quick_sort_random(arr):
    if len(arr) <= 1:
        return arr

    indice_pivo = random.randint(0, len(arr) - 1)
    pivo = arr[indice_pivo]

    esquerda = [x for i, x in enumerate(arr) if x < pivo]
    meio = [x for x in arr if x == pivo]
    direita = [x for i, x in enumerate(arr) if x > pivo]

    return quick_sort_random(esquerda) + meio + quick_sort_random(direita)

# Execução e Tempo
if vetor:

    inicio = time.perf_counter()
    vetor_ordenado = quick_sort_random(vetor)
    fim = time.perf_counter()

    tempo_total = fim - inicio

    print(f"Tempo de execução: {tempo_total:.6f} segundos")
    print(vetor_ordenado[:10])
else:
    print("Não foi possível processar o vetor.")

Tempo de execução: 0.032128 segundos
[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]


###Tempo de Execução:

Tentativa 1 - Tempo de execução: 0.030606 segundos  
Tentativa 2 - Tempo de execução: 0.031716 segundos  
Tentativa 3 - Tempo de execução: 0.034498 segundos  
Tentativa 4 - Tempo de execução: 0.032128 segundos  

##Vetor em ordem decrescente com 100000 elementos:

In [39]:
import requests
import time
import random

# Importar dados
url = "https://raw.githubusercontent.com/anabiarochar/Mestrado_AlgoritmosProgramacao/main/Projeto1/Vetores/vetor_100000_decrescente.txt"

try:
    response = requests.get(url)
    content = response.text
    # Ler arquivo como vetor separado por vírgula
    vetor = [int(x.strip()) for x in content.split(',') if x.strip()]
except Exception as e:
    print(f"Erro ao carregar o arquivo: {e}")

# Método QuickSort com pivô aleatório
def quick_sort_random(arr):
    if len(arr) <= 1:
        return arr

    indice_pivo = random.randint(0, len(arr) - 1)
    pivo = arr[indice_pivo]

    esquerda = [x for i, x in enumerate(arr) if x < pivo]
    meio = [x for x in arr if x == pivo]
    direita = [x for i, x in enumerate(arr) if x > pivo]

    return quick_sort_random(esquerda) + meio + quick_sort_random(direita)

# Execução e Tempo
if vetor:

    inicio = time.perf_counter()
    vetor_ordenado = quick_sort_random(vetor)
    fim = time.perf_counter()

    tempo_total = fim - inicio

    print(f"Tempo de execução: {tempo_total:.6f} segundos")
    print(vetor_ordenado[:10])
else:
    print("Não foi possível processar o vetor.")

Tempo de execução: 0.751031 segundos
[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]


###Tempo de Execução:

Tentativa 1 - Tempo de execução: 0.407084 segundos  
Tentativa 2 - Tempo de execução: 0.512516 segundos  
Tentativa 3 - Tempo de execução: 0.417885 segundos  
Tentativa 4 - Tempo de execução: 0.751031 segundos  

##Resumo dos Tempos de Execução

| Caso | n | Tentativa 1 | Tentativa 2 | Tentativa 3 | Tentativa 4 | Média |
|---|---:|---:|---:|---:|---:|---:|
| Aleatório | 1.000 | 0.003436 | 0.002381 | 0.002268 | 0.002113 | 0.002550 |
| Aleatório | 10.000 | 0.030176 | 0.064724 | 0.034632 | 0.032041 | 0.040393 |
| Aleatório | 100.000 | 0.860136 | 0.465810 | 0.523734 | 0.452698 | 0.575595 |
| Crescente | 1.000 | 0.007867 | 0.003974 | 0.002431 | 0.007712 | 0.005496 |
| Crescente | 10.000 | 0.030033 | 0.031734 | 0.032171 | 0.059693 | 0.038408 |
| Crescente | 100.000 | 0.366315 | 0.385219 | 0.404614 | 0.425225 | 0.395343 |
| Decrescente | 1.000 | 0.004115 | 0.002471 | 0.004370 | 0.002422 | 0.003345 |
| Decrescente | 10.000 | 0.030606 | 0.031716 | 0.034498 | 0.032128 | 0.032237 |
| Decrescente | 100.000 | 0.407084 | 0.512516 | 0.417885 | 0.751031 | 0.522129 |